### ISSUE TITLE
Performance of PER codec

### ISSUE THREAD

**Issue by dudycz** [librasn/rasn#244]
Hi. I have been playing with two most popular PER codecs: `rasn` and `asn1-codecs` and made some benchmark comparing their performance. One thing I have noticed is that rasn can be ~10x slower in encoding in some complex cases:

| Codec        | Encoding (µs)  | Decoding (µs)  |
|--------------|----------------|----------------|
| rasn         | 1916.7         | 826.8          |
| asn1-codecs  | 208            | 145.5          |

I looked into flamegraphs and callgrinds but I couldn't figure out what contributes to this big difference. If you're interested I could try to collect some and attach here.

Link to repo with benchmark: https://github.com/dudycz/asn1_codecs_bench

**Comment by XAMPPRocky** [comment]
Thank you for your issue! Would you be able to attach the flamegraphs? That would be very helpful. In general the initial version I made prioritised correctness over performance so there's likely a lot of places where it could be more efficient. I already know that I haven't added the empty struct optimisation that was added to BER to PER.

**Comment by dudycz** [comment]
I had to encode sample.asn 256 times to get something recorded in flamegraph. Here it is:
![rasn](https://github.com/librasn/rasn/assets/15647871/38bf1fc4-a6c1-4354-a1cb-3bc65adc94ca)

**Comment by Nicceboy** [comment]
Seems like your test bench uses only integers. `asn1-codecs` crate seems to use `i128` internally for Integer type while `rasn` uses [BigInts](https://docs.rs/crate/num-bigint/latest). I would guess that this plays big factor at least. 

I think there was a plan to make inner implementation of integer types as feature, so that it would be possible to select different internal type for performance reasons if very big numbers are not required.

**Comment by XAMPPRocky** [comment]
Yeah, that's still a todo, I should write up an issue giving details in case someone else who has more time is interested.

**Comment by Nicceboy** [comment]
> Yeah, that's still a todo, I should write up an issue giving details in case someone else who has more time is interested.

I think this is rather important optimisation in general and should not take too much work to implement, if we have, for example, just the i128 and BigInts as initial options. If you have time to write what you had on your mind, maybe I can try to do it.

**Comment by repnop** [comment]
@dudycz since #256 got merged a little while ago, are you able to rerun your benchmark, at least for the decoding side of things? it should be quite a bit better now.

**Comment by dudycz** [comment]
Yes, I can confirm big improvement in decoding time (on my "bench" setup I observed decrease from 457us to 341us!). Good job!

**Comment by dudycz** [comment]
Meanwhile I had updated my benchmark with third uper codec - asn1rs. 

| Codec        | Encoding (µs)  | Decoding (µs)  |
|--------------|----------------|----------------|
| rasn         | 1759           | 259            |
| asn1-codecs  | 187            | 122            |
| asn1rs       | 72             | 65             |

**Comment by XAMPPRocky** [comment]
Separately I think we should add continuous profiling to the CI and I've created an issue for that. Since this issue doesn't have a specific goal/end, I'm going to move this to a discussion.

https://github.com/librasn/rasn/issues/302
